src/util/rng.cpp

```diff
@@ -0,0 +1,62 @@
+/*********************                                                        */
+/*! \file rng.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Aina Niemetz
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2017 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief A Random Number Generator.
+ **
+ ** A random number generator, implements the xorshift* generator
+ ** (see S. Vigna, An experimental exploration of Marsaglia's xorshift
+ ** generators, scrambled. ACM Trans. Math. Softw. 42(4): 30:1-30:23, 2016).
+ **/
+
+#include "util/rng.h"
+
+#include <cfloat>
+#include "base/cvc4_assert.h"
+
+namespace CVC4 {
+
+uint64_t RNG::rand()
+{
+  /* xorshift* generator (see S. Vigna, An experimental exploration of
+   * Marsaglia's xorshift generators, scrambled. ACM Trans. Math. Softw.
+   * 42(4): 30:1-30:23, 2016). */
+  d_state ^= d_state >> 12;
+  d_state ^= d_state << 25;
+  d_state ^= d_state >> 27;
+  d_state *= UINT64_C(2685821657736338717);
+  return d_state;
+}
+
+uint64_t RNG::pick(uint64_t from, uint64_t to)
+{
+  Assert(from <= to);
+  Assert(to <= UINT64_MAX);
```

Comment:
Why not just say that `to < UINT64_MAX`? Right now the semantics change when `to == UINT64_MAX` because the range is not inclusive (but for all other inputs it is). The other option would be to only do the mod if `from == 0 && to == UINT64_MAX` is `false` (if it is `true`, just return the value from `rand()` directly).

Reply:
Good catch!